File: backend/internal/model/tournament.go
package model

import (
	"time"

	"github.com/google/uuid"
)

type Tournament struct {
	ID           uuid.UUID               `gorm:"type:uuid;primary_key;default:gen_random_uuid()" json:"id"`
	Name         string                  `gorm:"type:varchar(200);not null" json:"name"`
	MatchType    string                  `gorm:"type:varchar(10);not null" json:"match_type"` // "1v1" | "2v2"
	Status       string                  `gorm:"type:varchar(20);default:'active'" json:"status"` // active | completed
	AffectsScore bool                    `gorm:"default:true" json:"affects_score"`
	EntryFee     int                     `gorm:"default:0" json:"entry_fee"`
	CreatedAt    time.Time               `json:"created_at"`
	UpdatedAt    time.Time               `json:"updated_at"`
	Participants []TournamentParticipant `gorm:"foreignKey:TournamentID;constraint:OnDelete:CASCADE" json:"participants,omitempty"`
	Matches      []TournamentMatch       `gorm:"foreignKey:TournamentID;constraint:OnDelete:CASCADE" json:"matches,omitempty"`
}

func (Tournament) TableName() string { return "tournaments" }

type TournamentParticipant struct {
	ID                   uuid.UUID `gorm:"type:uuid;primary_key;default:gen_random_uuid()" json:"id"`
	TournamentID         uuid.UUID `gorm:"type:uuid;not null" json:"tournament_id"`
	UserID               uuid.UUID `gorm:"type:uuid;not null" json:"user_id"`
	TierSnapshot         string    `gorm:"type:varchar(10);default:'normal'" json:"tier_snapshot"`
	HandicapRateSnapshot float64   `gorm:"default:0.0" json:"handicap_rate_snapshot"`
	User                 User      `gorm:"foreignKey:UserID" json:"user,omitempty"`
}

func (TournamentParticipant) TableName() string { return "tournament_participants" }

type TournamentMatch struct {
	ID             uuid.UUID  `gorm:"type:uuid;primary_key;default:gen_random_uuid()" json:"id"`
	TournamentID   uuid.UUID  `gorm:"type:uuid;not null" json:"tournament_id"`
	Round          int        `gorm:"not null" json:"round"`
	MatchOrder     int        `gorm:"not null" json:"match_order"`
	Team1Player1ID uuid.UUID  `gorm:"type:uuid;not null" json:"team1_player1_id"`
	Team1Player2ID *uuid.UUID `gorm:"type:uuid" json:"team1_player2_id,omitempty"`
	Team2Player1ID uuid.UUID  `gorm:"type:uuid;not null" json:"team2_player1_id"`
	Team2Player2ID *uuid.UUID `gorm:"type:uuid" json:"team2_player2_id,omitempty"`
	HandicapTeam1  float64    `gorm:"default:0.0" json:"handicap_team1"`
	HandicapTeam2  float64    `gorm:"default:0.0" json:"handicap_team2"`
	Status         string     `gorm:"type:varchar(20);default:'pending'" json:"status"` // pending | completed
	ActualScore1   *int       `json:"actual_score1,omitempty"`
	ActualScore2   *int       `json:"actual_score2,omitempty"`
	EffectiveWinner int       `gorm:"default:0" json:"effective_winner"` // 0=draw/pending, 1=team1, 2=team2
	MatchID        *uuid.UUID `gorm:"type:uuid" json:"match_id,omitempty"`
	CreatedAt      time.Time  `json:"created_at"`
	UpdatedAt      time.Time  `json:"updated_at"`
	// Preloaded relations
	Team1Player1 *User `gorm:"-" json:"team1_player1,omitempty"`
	Team1Player2 *User `gorm:"-" json:"team1_player2,omitempty"`
	Team2Player1 *User `gorm:"-" json:"team2_player1,omitempty"`
	Team2Player2 *User `gorm:"-" json:"team2_player2,omitempty"`
}

func (TournamentMatch) TableName() string { return "tournament_matches" }

File: frontend/src/types/tournament.ts
import type { User } from './user'
import type { MatchType } from './match'

export type TournamentStatus = 'active' | 'completed'
export type TournamentFormat = 'classic' | 'round_robin_top4'
export type TournamentMatchStage = 'group' | 'semi' | 'final' | 'third_place'

export interface TournamentParticipant {
  id: string
  tournament_id: string
  user_id: string
  tier_snapshot: string
  handicap_rate_snapshot: number
  user: User
}

export interface TournamentTeam {
  id: string
  tournament_id: string
  player1_id: string
  player2_id: string
  player1_handicap_snapshot: number
  player2_handicap_snapshot: number
  player1: User
  player2: User
}

export interface TournamentMatch {
  id: string
  tournament_id: string
  round: number
  match_order: number
  stage: TournamentMatchStage
  team1_team_id?: string
  team2_team_id?: string
  team1_player1_id: string
  team1_player2_id?: string
  team2_player1_id: string
  team2_player2_id?: string
  handicap_team1: number
  handicap_team2: number
  status: 'pending' | 'completed'
  actual_score1?: number
  actual_score2?: number
  effective_winner: number // 0=draw, 1=team1, 2=team2
  match_id?: string
}

export interface Tournament {
  id: string
  name: string
  match_type: MatchType
  format: TournamentFormat
  status: TournamentStatus
  affects_score: boolean
  entry_fee: number
  knockout_size: number // 2 = final only; 4 = semis+final+3rd
  champion_team_id?: string
  champion_team?: TournamentTeam
  created_at: string
  updated_at: string
  participants: TournamentParticipant[]
  teams: TournamentTeam[]
  matches: TournamentMatch[]
  standings?: TeamStanding[]
}

export interface TeamInputEntry {
  player1_id: string
  player2_id: string
}

export interface CreateTournamentRequest {
  name: string
  match_type?: MatchType
  format?: TournamentFormat
  player_ids?: string[]
  teams?: TeamInputEntry[]
  affects_score: boolean
  entry_fee: number
  knockout_size?: number // 2 or 4; only relevant for round_robin_top4
}

export interface RecordMatchResultRequest {
  actual_score1: number
  actual_score2: number
  recorded_by: string
}

// Per-player standing (classic format, computed on frontend)
export interface TournamentStanding {
  user: User
  wins: number
  draws: number
  losses: number
  goals_for: number
  goals_against: number
  points: number
}

// Per-team standing (round_robin_top4, computed on backend)
export interface TeamStanding {
  team_id: string
  player1: User
  player2: User
  played: number
  won: number
  drawn: number
  lost: number
  gf: number
  ga: number
  gd: number
  points: number
  seed: number // 1-4 = qualified; 0 = eliminated
}
